README.md
# SilverStripe Foxy Feed Parser

An add-on module to parse the Foxy.io XML datafeed.

[![Build Status](https://travis-ci.org/dynamic/silverstripe-foxy-feed-parser.svg?branch=master)](https://travis-ci.org/dynamic/silverstripe-foxy-feed-parser)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/dynamic/silverstripe-foxy-feed-parser/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/dynamic/silverstripe-foxy-feed-parser/?branch=master)
[![Build Status](https://scrutinizer-ci.com/g/dynamic/silverstripe-foxy-feed-parser/badges/build.png?b=master)](https://scrutinizer-ci.com/g/dynamic/silverstripe-foxy-feed-parser/build-status/master)
[![codecov](https://codecov.io/gh/dynamic/silverstripe-foxy-feed-parser/branch/master/graph/badge.svg)](https://codecov.io/gh/dynamic/silverstripe-foxy-feed-parser)

[![Latest Stable Version](https://poser.pugx.org/dynamic/silverstripe-foxy-feed-parser/v/stable)](https://packagist.org/packages/dynamic/silverstripe-foxy-feed-parser)
[![Total Downloads](https://poser.pugx.org/dynamic/silverstripe-foxy-feed-parser/downloads)](https://packagist.org/packages/dynamic/silverstripe-foxy-feed-parser)
[![Latest Unstable Version](https://poser.pugx.org/dynamic/silverstripe-foxy-feed-parser/v/unstable)](https://packagist.org/packages/dynamic/silverstripe-foxy-feed-parser)
[![License](https://poser.pugx.org/dynamic/silverstripe-foxy-feed-parser/license)](https://packagist.org/packages/dynamic/silverstripe-foxy-feed-parser)


## Requirements

* SilverStripe ^4.0
* SilverStripe Foxy ^1.0

## Installation

```
composer require dynamic/silverstripe-foxy-feed-parser
```

## License

See [License](license.md)

## Maintainers
*  [Dynamic](http://www.dynamicagency.com) (<dev@dynamicagency.com>)
 
## Bugtracker
Bugs are tracked in the issues section of this repository. Before submitting an issue please read over 
existing issues to ensure yours is unique. 
 
If the issue does look like a new bug:
 
 - Create a new issue
 - Describe the steps required to reproduce your issue, and the expected outcome. Unit tests, screenshots 
 and screencasts can help here.
 - Describe your environment as detailed as possible: SilverStripe version, Browser, PHP version, 
 Operating System, any installed SilverStripe modules.
 
Please report security issues to the module maintainers directly. Please don't file security issues in the bugtracker.
 
## Development and contribution
If you would like to make contributions to the module please ensure you raise a pull request and discuss with the module maintainers.


CHANGELOG.md
# Changelog

## [1.1.0](https://github.com/dynamic/silverstripe-foxy-feed-parser/tree/1.1.0) (2020-07-18)

[Full Changelog](https://github.com/dynamic/silverstripe-foxy-feed-parser/compare/1.0.2...1.1.0)

**Merged pull requests:**

- REFACTOR replace options with variations in tests [\#16](https://github.com/dynamic/silverstripe-foxy-feed-parser/pull/16) ([jsirish](https://github.com/jsirish))

## [1.0.2](https://github.com/dynamic/silverstripe-foxy-feed-parser/tree/1.0.2) (2020-02-21)

[Full Changelog](https://github.com/dynamic/silverstripe-foxy-feed-parser/compare/1.0.1...1.0.2)

**Merged pull requests:**

- Added custom field mappings to transaction data [\#15](https://github.com/dynamic/silverstripe-foxy-feed-parser/pull/15) ([mak001](https://github.com/mak001))

## [1.0.1](https://github.com/dynamic/silverstripe-foxy-feed-parser/tree/1.0.1) (2019-11-08)

[Full Changelog](https://github.com/dynamic/silverstripe-foxy-feed-parser/compare/1.0.0...1.0.1)

**Merged pull requests:**

- Added fields for coupon mapping [\#14](https://github.com/dynamic/silverstripe-foxy-feed-parser/pull/14) ([mak001](https://github.com/mak001))
- BUGFIX Travis phpcs config [\#13](https://github.com/dynamic/silverstripe-foxy-feed-parser/pull/13) ([jsirish](https://github.com/jsirish))

## [1.0.0](https://github.com/dynamic/silverstripe-foxy-feed-parser/tree/1.0.0) (2019-10-30)

[Full Changelog](https://github.com/dynamic/silverstripe-foxy-feed-parser/compare/5f69af1715c3c53afefcd95e3be01b76c2a94140...1.0.0)

**Fixed bugs:**

- BUG incorrect variable usage in Transaction::getObject [\#3](https://github.com/dynamic/silverstripe-foxy-feed-parser/issues/3)

**Merged pull requests:**

- CHANGELOG 1.0.0 [\#12](https://github.com/dynamic/silverstripe-foxy-feed-parser/pull/12) ([jsirish](https://github.com/jsirish))
- Fixed discount mapping [\#11](https://github.com/dynamic/silverstripe-foxy-feed-parser/pull/11) ([mak001](https://github.com/mak001))
- Jenkins setup [\#9](https://github.com/dynamic/silverstripe-foxy-feed-parser/pull/9) ([jsirish](https://github.com/jsirish))
- BUGFIX FoxyController - processFoxyRequest - urlencode data before ex… [\#8](https://github.com/dynamic/silverstripe-foxy-feed-parser/pull/8) ([jsirish](https://github.com/jsirish))
- BUGFIX price of product not recording [\#7](https://github.com/dynamic/silverstripe-foxy-feed-parser/pull/7) ([muskie9](https://github.com/muskie9))
- ENHANCEMENT allow for retrieving encrypted data [\#5](https://github.com/dynamic/silverstripe-foxy-feed-parser/pull/5) ([muskie9](https://github.com/muskie9))
- BUGFIX use alternate data variable for getObject [\#4](https://github.com/dynamic/silverstripe-foxy-feed-parser/pull/4) ([muskie9](https://github.com/muskie9))
- REFACTOR remove legacy parsing code [\#2](https://github.com/dynamic/silverstripe-foxy-feed-parser/pull/2) ([muskie9](https://github.com/muskie9))
- ADD Transaction parser [\#1](https://github.com/dynamic/silverstripe-foxy-feed-parser/pull/1) ([muskie9](https://github.com/muskie9))
- BUGFIX test controller not working properly [\#6](https://github.com/dynamic/silverstripe-foxy-feed-parser/pull/6) ([muskie9](https://github.com/muskie9))



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


tests/Product/TestProduct.php
<?php

namespace Dynamic\Foxy\Parser\Tests\Product;

use Dynamic\Foxy\Extension\Shippable;
use SilverStripe\CMS\Model\SiteTree;
use SilverStripe\Dev\TestOnly;

/**
 * Class TestProduct
 * @package Dynamic\Foxy\Parser\Tests\Product
 */
class TestProduct extends SiteTree implements TestOnly
{
    /**
     * @var array
     */
    private static $extensions = [
        Shippable::class,
    ];
}


tests/Foxy/TransactionTest.php
<?php

namespace Dynamic\Foxy\Parser\Tests\Foxy;

use Dynamic\Foxy\Inventory\Test\TestOnly\Page\TestProduct;
use Dynamic\Foxy\Model\FoxyHelper;
use Dynamic\Foxy\Model\Variation;
use Dynamic\Foxy\Parser\Foxy\Transaction;
use Dynamic\Foxy\Parser\Tests\Controller\DataTestController;
use Dynamic\Foxy\Parser\Tests\FoxyXMLFeedFactory;
use Dynamic\Foxy\Parser\Tests\Product\FoxyFeedTestProduct;
use SilverStripe\Core\Config\Config;
use SilverStripe\Dev\FunctionalTest;
use SilverStripe\ORM\ArrayList;
use SilverStripe\View\ArrayData;

/**
 * Class TransactionTest
 * @package Dynamic\Foxy\Parser\Tests\Foxy
 */
class TransactionTest extends FunctionalTest
{
    /**
     * @var string
     */
    protected static $fixture_file = '../fixtures.yml';

    /**
     * @var array
     */
    protected static $extra_dataobjects = [
        FoxyFeedTestProduct::class,
    ];

    /**
     *
     */
    protected function setUp()
    {
        parent::setUp();

        Config::modify()->set(Variation::class, 'has_one', ['Product' => FoxyFeedTestProduct::class]);
    }

    /**
     * @throws \SilverStripe\Security\PasswordEncryptor_NotFoundException
     */
    protected static function get_foxy_data()
    {
        $password = 'password';
        $hashType = 'sha1_v2.4';
        $salt = 'faGgWXUTdZ7i42lpA6cljzKeGBeUwShBSNHECwsJmt';

        return [
            "TransactionDate" => strtotime("now"),
            "OrderID" => rand(0, 5000),
            "Email" => FoxyXMLFeedFactory::generate_email(),
            "Password" => FoxyXMLFeedFactory::get_hashed_password($hashType, $password, $salt),
            'Salt' => $salt,
            'HashType' => $hashType,
            "OrderDetails" => ArrayData::create([
                'Title' => 'foo',
                'Price' => 20.00,
                'Quantity' => 1,
                'Weight' => 0.1,
                'DeliveryType' => 'shipped',
                'CategoryDescription' => 'Default cateogry',
                'CategoryCode' => 'DEFAULT',
                'Options' => [
                    [
                        'Name' => 'color',
                        'OptionValue' => 'blue',
                        'PriceMod' => '',
                        'WeightMod' => '',
                    ],
                    [
                        'Name' => 'product_id',
                        'OptionValue' => 5,
                        'PriceMod' => '',
                        'WeightMod' => '',
                    ],
                ],
            ]),
        ];
    }

    /**
     * @throws \Exception
     */
    public static function setUpBeforeClass()
    {
        parent::setUpBeforeClass();

        Config::modify()->set(FoxyHelper::class, 'secret', 'abc123');
        Config::modify()->merge(DataTestController::class, 'allowed_actions', ['index', 'xml', 'encryptedXML']);
        Config::modify()->set(DataTestController::class, 'data', static::get_foxy_data());
        Config::modify()->set(DataTestController::class, 'run_config_update', false);
    }

    /**
     *
     */
    public function testTransactionConstruct()
    {
        $response = $this->get('/foxytest/xml');

        if (PHP_VERSION_ID < 70100) {
            $this->markTestSkipped('PHPUnit_Framework_Error_Warning exception is thrown for legacy PHP versions only');
        } else {
            $this->expectException(\ArgumentCountError::class);
        }

        Transaction::create();

        $this->assertInstanceOf(Transaction::class, Transaction::create($response));
    }

    /**
     * @throws \SilverStripe\ORM\ValidationException
     * @throws \SilverStripe\Security\PasswordEncryptor_NotFoundException
     */
    public function testGetParsedTransactionData()
    {
        $response = DataTestController::singleton()->encryptedXML();

        $this->assertInstanceOf(ArrayData::class, Transaction::create($response)->getParsedTransactionData());
    }

    /**
     * @throws \SilverStripe\ORM\ValidationException
     * @throws \SilverStripe\Security\PasswordEncryptor_NotFoundException
     */
    public function testAttributesExist()
    {
        $response = DataTestController::singleton()->encryptedXML();
        $transaction = Transaction::create($response)->getParsedTransactionData();

        $this->assertInstanceOf(ArrayList::class, $transaction->products);
        $this->assertInstanceOf(ArrayList::class, $transaction->discounts);
    }
}


tests/Factory/FoxyXMLFeedFactory.php
<?php

namespace Dynamic\Foxy\Parser\Tests;

use Dynamic\Foxy\Model\FoxyHelper;
use SilverStripe\Control\Controller;
use SilverStripe\Core\Config\Configurable;
use SilverStripe\Core\Injector\Injectable;
use SilverStripe\Security\Member;
use SilverStripe\Security\PasswordEncryptor;

class FoxyXMLFeedFactory
{
    use Configurable;
    use Injectable;

    /**
     * @var Controller
     */
    private $controller;

    /**
     * @var array
     */
    private $feed_array = [];


    /**
     * @var array
     */
    private static $feed_array_config = [];

    /**
     * FoxyXMLFeedFactory constructor.
     * @param Controller $controller
     * @param array $feedData
     */
    public function __construct(Controller $controller, $feedData = [])
    {
        $this->setController($controller);

        if (is_array($feedData) && !empty($feedData)) {
            $this->setFeedArray($feedData);
        }
    }

    /**
     * @return string
     */
    public static function generate_email()
    {
        $emails = Member::get()->filter([
            'Email:EndsWith' => '@example.com',
        ])->column('Email');

        if ($emails && count($emails)) {
            $email = $emails[count($emails) - 1];

            return preg_replace_callback(
                "|(\d+)|",
                function ($mathces) {
                    return ++$mathces[1];
                },
                $email
            );
        }

        return 'example0@example.com';
    }

    /**
     * @param string $algorithm
     * @param string $password
     * @param string $salt
     * @return String
     * @throws \SilverStripe\Security\PasswordEncryptor_NotFoundException
     */
    public static function get_hashed_password($algorithm, $password, $salt)
    {
        $encryptor = PasswordEncryptor::create_for_algorithm($algorithm);

        return $encryptor->encrypt($password, $salt);
    }

    /**
     * @param $controller
     * @return $this
     */
    protected function setController($controller)
    {
        $this->controller = $controller;

        return $this;
    }

    /**
     * @return Controller
     */
    protected function getController()
    {
        return $this->controller;
    }

    /**
     * @param array $feedData
     * @return $this
     */
    public function setFeedArray($feedData = [])
    {
        if (is_array($feedData) && !empty($feedData)) {
            $this->feed_array = $feedData;
        } else {
            $this->feed_array = $this->config()->get('feed_array_config');
        }

        return $this;
    }

    /**
     * @return array
     */
    public function getFeedArray()
    {
        return $this->feed_array;
    }

    /**
     * @return mixed
     */
    public function getXML()
    {
        $xml = Controller::curr()->renderWith('TestData', $this->getFeedArray());

        return $xml->RAW();
    }

    /**
     * @return string
     * @throws \SilverStripe\ORM\ValidationException
     * @throws \SilverStripe\Security\PasswordEncryptor_NotFoundException
     */
    public function encryptedXML()
    {
        $helper = FoxyHelper::singleton();

        return \rc4crypt::encrypt($helper->config()->get('secret'), $this->getXML());
    }
}


src/Foxy/Transaction.php
<?php

namespace Dynamic\Foxy\Parser\Foxy;

use Dynamic\Foxy\Model\FoxyHelper;
use SilverStripe\Core\Config\Configurable;
use SilverStripe\Core\Injector\Injectable;
use SilverStripe\ORM\ArrayList;
use SilverStripe\ORM\ValidationException;
use SilverStripe\View\ArrayData;

/**
 * Class Transaction
 * @package Dynamic\Foxy\Parser\Foxy
 */
class Transaction
{
    use Injectable;
    use Configurable;

    /**
     * @var string Encrypted response from Foxy
     */
    private $encrypted_data;

    /**
     * @var "Foxy.io transaction xml record"
     */
    private $transaction;

    /**
     * @var ArrayData
     */
    private $transaction_data;

    /**
     * @var ArrayList
     */
    private $discount_data;

    /**
     * @var ArrayList
     */
    private $product_data;

    /**
     * @var ArrayList
     */
    private $custom_fields_data;

    /**
     * Transaction constructor.
     * @param $data string encrypted foxy response data from the xml data feed
     * @throws ValidationException
     */
    public function __construct($data)
    {
        $this->setEncryptedData($data);
        $this->setTransaction($data);
    }

    /**
     * @param $data
     * @return $this
     */
    public function setEncryptedData($data)
    {
        $this->encrypted_data = $data;

        return $this;
    }

    /**
     * @return string
     */
    public function getEncryptedData()
    {
        return $this->encrypted_data;
    }

    /**
     * Set the decrypted transaction data to use.
     *
     * @param $data
     * @return $this
     * @throws ValidationException
     */
    public function setTransaction($data)
    {
        $decryptedData = $this->getDecryptedData($data);

        foreach ($decryptedData->transactions->transaction as $transaction) {
            if ($this->hasID($transaction)) {
                $this->transaction = $transaction;
                break;
            }
        }

        if (!$this->transaction) {
            $this->transaction = false;
        }

        return $this;
    }

    /**
     * Check if there is a transaction ID for the given decrypted transaction data.
     *
     * @param $transaction
     * @return bool
     */
    protected function hasID($transaction)
    {
        return (int)$transaction->id > 0;
    }

    /**
     * Return the decrypted transaction xml data.
     *
     * @return mixed
     */
    protected function getTransaction()
    {
        return $this->transaction;
    }

    /**
     * Return an ArrayData containing transaction, products and discounts data.
     *
     * @return ArrayData
     */
    public function getParsedTransactionData()
    {
        /**
         *
         */
        return ArrayData::create([
            'transaction' => $this->getTransactionData(),
            'products' => $this->getProductData(),
            'discounts' => $this->getDiscountData(),
            'custom_fields' => $this->getCustomFieldsData(),
        ]);
    }

    /**
     * Set the base transaction data in an ArrayData object.
     *
     * @return $this
     */
    private function setTransactionData()
    {
        $this->transaction_data = $this->getObject(
            $this->getTransaction(),
            $this->config()->get('transaction_mapping')
        );

        return $this;
    }

    /**
     * Return base transaction data.
     *
     * @return ArrayData
     */
    protected function getTransactionData()
    {
        if (!$this->transaction_data instanceof ArrayData) {
            $this->setTransactionData();
        }

        return $this->transaction_data;
    }

    /**
     * Return discounts data from Foxy.io xml data feed if any.
     *
     * @return ArrayList
     */
    protected function getDiscountData()
    {
        if (!$this->discount_data instanceof ArrayList) {
            $this->setDiscountData();
        }

        return $this->discount_data;
    }

    /**
     * Set an ArrayList with possible discount data as ArrayData objects.
     *
     * @return $this
     */
    protected function setDiscountData()
    {
        $discounts = $this->getTransaction()->discounts;
        $discountsList = ArrayList::create();

        foreach ($discounts as $discount) {
            $discountsList->push($this->getObject($discount->discount, $this->config()->get('discounts_mapping')));
        }

        $this->discount_data = $discountsList;

        return $this;
    }

    /**
     * Return transaction_details data from Foxy.io xml data feed if any.
     *
     * @return ArrayList
     */
    protected function getProductData()
    {
        if (!$this->product_data instanceof ArrayList) {
            $this->setProductData();
        }

        return $this->product_data;
    }

    /**
     * Set an ArrayList with possible transaction_detail data as ArrayData objects.
     *
     * @return $this
     */
    protected function setProductData()
    {
        $details = $this->getTransaction()->transaction_details->transaction_detail;
        $detailsList = ArrayList::create();

        foreach ($details as $detail) {
            $product = $this->getObject($detail, $this->config()->get('transaction_detail_mapping'));

            $product->transaction_detail_options =
                $this->getProductOptions($detail->transaction_detail_options->transaction_detail_option);

            $detailsList->push($product);
        }

        $this->product_data = $detailsList;

        return $this;
    }

    /**
     * Returns an ArrayList containing possible transaction_detail_option data as ArrayData objects.
     *
     * @param $data
     * @return ArrayList
     */
    protected function getProductOptions($data)
    {
        $options = ArrayList::create();

        foreach ($data as $option) {
            $options->push($this->getObject($option, $this->config()->get('transaction_detail_option_mapping')));
        }

        return $options;
    }

    /**
     * @return ArrayList
     */
    protected function getCustomFieldsData()
    {
        if (!$this->custom_fields_data instanceof ArrayList) {
            $this->setCustomFieldsData();
        }

        return $this->custom_fields_data;
    }

    /**
     * @return $this
     */
    protected function setCustomFieldsData()
    {
        $customFields = $this->getTransaction()->custom_fields;
        $customFieldsList = ArrayList::create();

        foreach ($customFields as $customFieldArray) {
            foreach ($customFieldArray as $customField) {
                $customFieldsList->push($this->getObject(
                    $customField,
                    $this->config()->get('custom_fields_mapping')
                ));
            }
        }

        $this->custom_fields_data = $customFieldsList;
        return $this;
    }

    /**
     * Returns an ArrayData object based on the given iterable data and a key/val config array. Used
     * to type hint data from the Foxy.io xml data feed.
     *
     * @param $data
     * @param array $config
     * @return ArrayData
     */
    protected function getObject($data, $config = [])
    {
        $dataArray = [];

        foreach ($config as $name => $type) {
            switch ($type) {
                case 'int':
                    $dataArray[$name] = (int)$data->{$name};
                    break;
                case 'float':
                    $dataArray[$name] = (float)$data->{$name};
                    break;
                case 'bool':
                case 'boolean':
                    $dataArray[$name] = (bool)$data->{$name};
                    break;
                case 'string':
                default:
                    $dataArray[$name] = (string)$data->{$name};
                    break;
            }
        }

        return ArrayData::create($dataArray);
    }

    /**
     * @return bool
     */
    public function exists()
    {
        return $this->getTransaction() != false && !empty($this->getTransaction());
    }

    /**
     * @param $data
     * @return \SimpleXMLElement
     * @throws \SilverStripe\ORM\ValidationException
     */
    private function getDecryptedData($data)
    {
        $helper = new FoxyHelper();

        return new \SimpleXMLElement(\rc4crypt::decrypt($helper->config()->get('secret'), $data));
    }
}


src/Controller/FoxyController.php
<?php

namespace Dynamic\Foxy\Parser\Controller;

use Dynamic\Foxy\Parser\Foxy\Transaction;
use SilverStripe\Control\Controller;
use SilverStripe\Control\HTTPRequest;
use SilverStripe\ORM\ValidationException;

/**
 * Class FoxyController
 * @package Dynamic\Foxy\Parser\Controller
 */
class FoxyController extends Controller
{
    /**
     *
     */
    public const URLSEGMENT = 'foxy';

    /**
     * @var array
     */
    private static $allowed_actions = [
        'index',
    ];

    /**
     * @return string
     */
    public function getURLSegment()
    {
        return self::URLSEGMENT;
    }

    /**
     * @return string
     * @throws ValidationException
     */
    public function index()
    {
        $request = $this->getRequest();
        if ($request->postVar('FoxyData') || $request->postVar('FoxySubscriptionData')) {
            $this->processFoxyRequest($request);

            return 'foxy';
        }

        return 'No FoxyData or FoxySubscriptionData received.';
    }

    /**
     * Process a request after a transaction is completed via Foxy
     *
     * @param HTTPRequest $request
     * @throws ValidationException
     */
    protected function processFoxyRequest(HTTPRequest $request)
    {
        $encryptedData = $request->postVar('FoxyData')
            ? urldecode($request->postVar('FoxyData'))
            : urldecode($request->postVar('FoxySubscriptionData'));
        $this->parseFeedData($encryptedData);

        $encryptedData = urlencode($encryptedData);

        $this->extend('addIntegrations', $encryptedData);
    }

    /**
     * Parse the XML data feed from Foxy to a SimpleXMLElement object
     *
     * @param $encryptedData
     * @param $decryptedData
     * @throws ValidationException
     */
    private function parseFeedData($encryptedData)
    {
        $transaction = Transaction::create($encryptedData);

        $this->extend('doAdditionalParse', $transaction);
    }
}


src/Controller/FoxyDataGeneratorController.php
<?php

namespace Dynamic\Foxy\Parser\Controller;

use Dynamic\Foxy\Extension\Purchasable;
use Dynamic\Foxy\Extension\Shippable;
use Dynamic\Foxy\Model\FoxyHelper;
use GuzzleHttp\Client;
use SilverStripe\CMS\Model\SiteTree;
use SilverStripe\Control\Controller;
use SilverStripe\Control\Director;
use SilverStripe\Core\Config\Config;
use SilverStripe\Core\Injector\Injector;
use SilverStripe\Dev\DebugView;
use SilverStripe\ORM\ArrayList;
use SilverStripe\Security\Member;
use SilverStripe\Security\PasswordEncryptor;

/**
 * Class FoxyDataGeneratorController
 * @package Dynamic\Foxy\Parser\Controller
 */
class FoxyDataGeneratorController extends Controller
{
    /**
     * @var array
     */
    private static $allowed_actions = [
        'index',
        'xml',
        'encryptedXML',
    ];

    /**
     * @var array
     */
    private static $data = [
        "TransactionDate" => "now",
        "OrderID" => "auto",
        "Email" => "auto",
        "Password" => "password",
        "OrderDetails" => [],
    ];

    /**
     * @throws \GuzzleHttp\Exception\GuzzleException
     * @throws \SilverStripe\ORM\ValidationException
     * @throws \SilverStripe\Security\PasswordEncryptor_NotFoundException
     */
    public function index()
    {
        $rules = Director::config()->get('rules');
        $rule = array_search(FoxyController::class, $rules);
        $myURL = Director::absoluteBaseURL() . explode('//', $rule)[0];

        $helper = new FoxyHelper();
        $myKey = $helper->config()->get('secret');

        $XMLOutput_encrypted = $this->encryptedXML();
        $XMLOutput_encrypted = urlencode($XMLOutput_encrypted);

        $client = new Client();
        $response = $client->request('POST', $myURL, [
            'form_params' => ['FoxyData' => $XMLOutput_encrypted],
        ]);

        $configString = print_r(static::config()->get('data'), true);
        /** @var DebugView $view */
        $view = Injector::inst()->create(DebugView::class);
        echo $view->renderHeader();
        echo '<div class="info">';
        echo "<h2>Config:</h2><pre>$configString</pre>";
        if ($this->getRequest()->getVar('data')) {
            //echo "<h2>Data:</h2><pre>{$xml->HTML()}</pre>";
        }
        echo "<h2>Response:</h2><pre>" . $response->getBody() . "</pre>";
        echo '<p></p>';
        echo '</div>';
        echo $view->renderFooter();
    }

    /**
     * @throws \SilverStripe\Security\PasswordEncryptor_NotFoundException
     */
    private function updateConfig()
    {
        $data = Config::inst()->get(self::class, 'data');
        $transactionDate = $data['TransactionDate'];
        static::config()->merge('data', [
            'TransactionDate' => strtotime($transactionDate),
        ]);

        $orderID = $data['OrderID'];
        if ($orderID === 'auto' || $orderID < 1) {
            static::config()->merge('data', [
                'OrderID' => rand(0, 5000),
            ]);
        }

        $email = $data['Email'];
        if ($email === 'auto') {
            static::config()->merge('data', [
                'Email' => $this->generateEmail(),
            ]);
        }

        $orderDetails = $data['OrderDetails'];
        if ($orderDetails === null || count($orderDetails) === 0) {
            static::config()->merge('data', [
                'OrderDetails' => [
                    $this->generateOrderDetail(),
                ],
            ]);
        }

        if (!array_key_exists('Salt', $data)) {
            static::config()->merge('data', [
                'Salt' => 'faGgWXUTdZ7i42lpA6cljzKeGBeUwShBSNHECwsJmt',
            ]);
        }

        if (!array_key_exists('HashType', $data)) {
            static::config()->merge('data', [
                'HashType' => 'sha1_v2.4',
            ]);
        }

        $data = static::config()->get('data');
        if (!array_key_exists('HashedPassword', $data)) {
            $encryptor = PasswordEncryptor::create_for_algorithm($data['HashType']);
            static::config()->merge('data', [
                'HashedPassword' => $encryptor->encrypt($data['Password'], $data['Salt']),
            ]);
        }
    }

    /**
     * @return string
     */
    private function generateEmail()
    {
        $emails = Member::get()->filter([
            'Email:EndsWith' => '@example.com',
        ])->column('Email');

        if ($emails && count($emails)) {
            $email = $emails[count($emails) - 1];

            return preg_replace_callback(
                "|(\d+)|",
                function ($mathces) {
                    return ++$mathces[1];
                },
                $email
            );
        }

        return 'example0@example.com';
    }

    /**
     * @return array
     */
    private function generateOrderDetail()
    {
        return [
            'Title' => 'foo',
            'Price' => 20.00,
            'Quantity' => 1,
            'Weight' => 0.1,
            'DeliveryType' => 'shipped',
            'CategoryDescription' => 'Default cateogry',
            'CategoryCode' => 'DEFAULT',
            'Options' => ArrayList::create([
                [
                    'Name' => 'color',
                    'OptionValue' => 'blue',
                    'PriceMod' => '',
                    'WeightMod' => '',
                ],
                [
                    'Name' => 'product_id',
                    'OptionValue' => $this->getTestProduct(),
                    'PriceMod' => '',
                    'WeightMod' => '',
                ],
            ]),
        ];
    }

    /**
     * @return bool
     */
    public function getTestProduct()
    {
        $pages = SiteTree::get();
        foreach ($pages as $page) {
            if ($page->hasExtension(Purchasable::class) || $page->hasExtension(Shippable::class)) {
                return $page->ID;
            }
        }

        return false;
    }

    /**
     * @return mixed|string
     * @throws \SilverStripe\Security\PasswordEncryptor_NotFoundException
     */
    public function xml()
    {
        $this->updateConfig();

        $config = static::config()->get('data');

        $config['OrderDetails'] = ArrayList::create($config['OrderDetails']);

        $xml = $this->renderWith('TestData', $config);

        return $xml->RAW();
    }

    /**
     * @return string
     * @throws \SilverStripe\ORM\ValidationException
     * @throws \SilverStripe\Security\PasswordEncryptor_NotFoundException
     */
    public function encryptedXML()
    {
        $helper = new FoxyHelper();

        return \rc4crypt::encrypt($helper->config()->get('secret'), $this->xml());
    }
}
